src/mlpack/tests/recurrent_network_test.cpp

```diff
@@ -456,11 +457,11 @@ void ReberGrammarTestNetwork(const size_t hiddenSize = 4,
     const size_t outputSize = 7;
     const size_t inputSize = 7;
 
-    RNN<MeanSquaredError<>, RandomInitialization, Args...> model(5);
-    model.template Add<Linear<> >(inputSize, hiddenSize);
-    model.template Add<RecurrentLayerType>(hiddenSize, hiddenSize);
-    model.template Add<Linear<> >(hiddenSize, outputSize);
-    model.template Add<OutputLayerType>();
+    RNN<MeanSquaredError<> > model(5);
+    model.Add<Linear<> >(inputSize, hiddenSize);
+    model.Add<RecurrentLayerType>(hiddenSize, hiddenSize);
+    model.Add<Linear<> >(hiddenSize, outputSize);
+    model.Add<SigmoidLayer<> >();
```

Comment:
It should be possible to make the templatization work on gcc 5 and 6; this shouldn't be too much of an issue.  Maybe this is a C++ standard issue, so if you are interested in debugging further you might be able to reproduce the issue on your version of gcc by additionally making sure that the `-std=c++11` flag is given to the compiler.  I think this can be done by setting `CMAKE_CXX_FLAGS="-std=c++11"` when you configure.

Reply:
I tried those. Problem is the same code compiled with GCC-7 and Clang-5 but not with lower versions of GCC. 
I tried to add a variadic template parameter to the `BuildNetwork` function and add one more template parameter for the output layer type. The variadic parameter is necessary because we cannot add the same type in `LayerTypes` more than once.  
I will investigate some more the cause of the error. Something to do with template substitution. Perhaps a GCC bug, which cannot handle the variadic template properly. But I thought it was not necessary to hold up this work for that issue. I will create a new pull request if I find a more elegant solution.

Reply:
Agreed, no need to hold up this PR.  We can return to the issue later if you figure something out.  I think what you had written was valid C++11 and older versions should support it too.  It's possible it could be a compiler bug; we have encountered those before.